### DISCUSSION TITLE
Probable Bug: ClientResponseError 521 in random moments.

### DISCUSSION THREAD

**Discussion by sum117** [pocketbase/pocketbase#2586] [Q&A] [answered]
Good afternoon.

I'm having some issues. This error is popping up at random times, and I can't tell why. It's breaking the bot for some people, forcing me to restart it.

[Error is happening in this file in my project](https://github.com/sum117/ansenbot/blob/master/src/events/OnRoleplayChannelActivity.ts)

```sh
ClientResponseError 521: Something went wrong while processing your request.
    at new ClientResponseError (file:///app/node_modules/.pnpm/pocketbase@0.14.4/node_modules/pocketbase/dist/pocketbase.es.mjs:1:2306)
    at file:///app/node_modules/.pnpm/pocketbase@0.14.4/node_modules/pocketbase/dist/pocketbase.es.mjs:1:45357
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async OnRoleplayChannelActivity.processRoleplayingChannels (file:///app/build/src/events/OnRoleplayChannelActivity.js:121:21)
    at async OnRoleplayChannelActivity.main (file:///app/build/src/events/OnRoleplayChannelActivity.js:34:17)
    at async next (file:///app/node_modules/.pnpm/discordx@11.7.6_discord.js@14.11.0/node_modules/discordx/build/esm/decorators/classes/Method.js:61:23)
    at async file:///app/node_modules/.pnpm/discordx@11.7.6_discord.js@14.11.0/node_modules/discordx/build/esm/logic/metadata/MetadataStorage.js:418:29
    at async Promise.all (index 0)
    at async Client.<anonymous> (file:///app/node_modules/.pnpm/discordx@11.7.6_discord.js@14.11.0/node_modules/discordx/build/esm/logic/metadata/MetadataStorage.js:414:13) {
  url: 'https://pocketbase.sumserver.xyz/api/collections/channels/records/ymsy58ysk4nkr4g?expand=faction%2Crace%2Cskills%2Cstatus%2Ceffects%2Cplayer%2Cmemory%2Cspecs%2CdestinyMaiden%2CstartingSkills%2Cinventory.consumables.item%2Cinventory.spells.item%2Cinventory.equipments.item%2Citem%2Cconsumables(item)%2Cequipments(item)%2Cspells(item)%2Cbody.head.item%2Cbody.face.item%2Cbody.shoulders.item%2Cbody.chest.item%2Cbody.amulet.item%2Cbody.back.item%2Cbody.legs.item%2Cbody.feet.item%2Cbody.leftArm.item%2Cbody.rightArm.item%2Cbody.rings.item%2Cbody.spells.item%2Chead.item%2Cface.item%2Cshoulders.item%2Cchest.item%2Camulet.item%2Cback.item%2Clegs.item%2Cfeet.item%2CleftArm.item%2CrightArm.item%2Crings.item%2Cspells.item',
  status: 521,
  response: {},
  isAbort: false,
  originalError: null
}
file:///app/node_modules/.pnpm/pocketbase@0.14.4/node_modules/pocketbase/dist/pocketbase.es.mjs:1
```

**Comment by ganigeorgiev** [comment]
I don't think we explicitly return 521 error anywhere.

If you have a reverse proxy like nginx or cloudflare double check its configuration.

Also I recommend updating to the latest [v0.16.3 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.16.3) as it includes some fixes related to the default db connections options.

**Reply by sum117** [reply]
Will do.

**Reply by sum117** [reply]
Just checked. I'm using the latest docker image. Maybe the issue is with using docker? Idk. 

ghcr.io/muchobien/pocketbase   latest    50ad2f8264ad   3 days ago     55.3MB

EDIT: Nvm. The 0.16.3 release is brand new. Sorry.

**Reply by ganigeorgiev** [reply, marked as answer]
The 521 error is probably caused by something with the docker setup as we don't explicitly return this response code, but I'm not sure what exactly. Googleing the error code indicates that usually this is thrown by proxies when the server is down or unreachable.

There are too many things going on in your repo and I'm not sure how to reproduce it. You can try to monitor your container memory usage and see if it is too high. Also note that you are requesting too many things at once (**the above url in the stacktrace has 40+ nested expands just in a single request!**) and it may be a good idea to split it in several requests and/or lazy load only the content you need.
If you are memory constrained, you can also try to register the env variable [`GOMEMLIMIT=500MiB`](https://pkg.go.dev/runtime@master#hdr-Environment_Variables) (it is a soft-limit).

**Reply by sum117** [reply]
Thank you for taking a look into this!
